### ISSUE TITLE
Keyboard usage workflows

### ISSUE THREAD

**Issue by tuomas2** [AndBible/and-bible#1878]
Related: #60

And Bible is becoming more and more usable with devices with keyboard (Windows 11, Android emulators on PC/mac, bluetooth keyboards with tablet etc). With those devices it is not convenient to be touching the screen all the time. 

I would like to have shortcuts to help navigating with keyboard over the interface. Most common operations
should be doable so that I don't waste time on touching screen for them.

Window operations:

- [x] Switch from window to another (alt+up/down)
- [x] Enter window button bar mode (alt + w)
  - [x] Open window menu (Space)
  - [x] Open/restore minimized window (Enter)
  - [ ] Minimize window 

Others:

- [ ] Find feature should be more accessible without touching the screen
- [x] Enter bible verse / key selector (Alt+G)
- [ ] Go to a bible passage (by typing) 
- [ ] Switch to next/previous workspace
- [ ] Show workspace list, select and open workspace
- [x] Open main menu (Alt+M)
- [x] Open options menu (Alt+O)
- [ ] How to exit full screen?
- [ ] Focus toolbar buttons

Navigating on bibleview:

- [ ] Cursor to indicate verse when using keyboard
- [ ] 1-click verse selection popup with enter key
- [ ] add bookmark etc with keyboard
- [ ] study pad navigation: adding new note, closing note, navigation, "cursor".
- [ ] Go back (Alt + Left arrow (chrome shortcut) ?)
- [x] ESC to close modals
- [x] ESC to close edit modes
- [ ] Key to travel back in history
- [ ] Minimise current window
- [ ] Maximise current window
- [ ] Open current window menu
- [ ] Going down / up smoothly in view
- [ ] Going down / up verse by verse

Probably we need to make "keyboard mode" in which:

- [ ] We need to show some keyboard hints of non-obvious keys (like in 1-click bookmark modal etc)
- [ ] Need to have help screen for keyboard keys somewhere (if keyboard mode enabled)

Some parts can be done already for 4.0 (like adding some shortcut keys) but most of this can be done earliest in 4.1+.

**Comment by tuomas2** [comment]
Related: #1583

**Comment by tuomas2** [comment]
Hmm ESC key and android's back button seem to give same keycode. That's bad. I don't want to use ESC to travel back in history. 

 -> okay, I can distinquish it by KeyEvent.getSource

**Comment by agrogers** [comment]
It seems OK to me. ESC undoes last action.

**Comment by tuomas2** [comment]
You mean that you would want ESC to go back in history? I think it causes too many unintended back-presses. There must be key for going back in history too, of course.

**Comment by agrogers** [comment]
If I want to close a dialog I would go for ESC. And if ESC is Back then
that works nicely

**Comment by tuomas2** [comment]
Yes, ESC is "back" in all other ways (closing dialogs, modals etc), but not for going back in history.

**Comment by tuomas2** [comment]
There are more and more functional key bindings in 619. Please test and let me know how it feels.

**Comment by tuomas2** [comment]
> Going down / up smoothly in view
> Going down / up verse by verse

Webview has its own keyboard "link by link" navigation which seem to be enabled by default. It works sometimes "okay" but sometimes not so nicely.

**Comment by tuomas2** [comment]
Some docs about android keyboard navigation in general https://developer.android.com/training/keyboard-input/navigation

**Comment by tuomas2** [comment]
Perhaps we should do like this:
- handle tab and arrow keys like normally, so that each webview is "single focus point". 
  - Show focusing of bible views differently when not in bibleview mode (perhaps normal "grey" overlay).
- Then enter into bibleview mode by some key, like enter. 
  - After entering bibleview mode, you navigate inside webview, with more specific navigation that we can define. 
  - From bibleview, you can use alt+* bindings too, to quickly access items outside bibleview.
  - Show focus somehow ("current window indicators" could then change color?)
  - When we are not in text edit field, we can repurpose / handle also tab and arrow buttons.
- Key to exit bibleview mode (esc, if there is no modal open).

**Comment by agrogers** [comment]
Are you referring to navigating between, say, individual Strong's numbers?
Or moving the highlighted verse up and down? That sounds pretty complicated.

I haven't used a keyboard for almost 2 years - my chromebook is stuck in
Yangon (or rather I'm stuck in Australia). So I don't know how I would
want it to work.

But I like the idea of having two modes - one to navigate outside of
webview and one that is inside - did I understand correctly?

What about Alt + Up/Down arrow to move between windows and break out of
webview. That could give focus to the window itself (not the button). Could
identify that by making the corner highlights thicker. Hitting Enter could
then drop you into webview.

Or maybe I am missing your intentions altogether?

**Comment by tuomas2** [comment]
> Are you referring to navigating between, say, individual Strong's numbers? Or moving the highlighted verse up and down? That sounds pretty complicated. 

Which part of my text are you referring to? I'm missing something.

> I haven't used a keyboard for almost 2 years - my chromebook is stuck in Yangon (or rather I'm stuck in Australia). So I don't know how I would want it to work. But I like the idea of having two modes - one to navigate outside of webview and one that is inside - did I understand correctly?

I think you did understand at least somewhat correctly.

> What about Alt + Up/Down arrow to move between windows and break out of webview. That could give focus to the window itself (not the button). Could identify that by making the corner highlights thicker. Hitting Enter could then drop you into webview. Or maybe I am missing your intentions altogether?

I'll experiment something, let's review my ideas after I get something done :-P 

Do you think it's okay if these changes are included in prod builds, even though they might change (no "UI" changes)? I thought it's okay, because the "current" keyboard usage is completely unusable, so anything is improvement. And most people don't notice anything because they don't use keyboard.

**Comment by agrogers** [comment]
Yeah, I wouldn't have an issue with it being included in the production
build.

**Comment by timbze** [comment]
Wish I could test but seems Android Emulator is horrible with accepting keyboard shortcuts. In 1 mode I find only escape works, and in another mode only alt + up and alt + down work. Very weird

**Comment by tuomas2** [comment]
I agree. I have to work continuously with my tablet (+ bt keyboard) when doing this.

**Comment by tuomas2** [comment]
notes from church 

> AB: keyboardilla ikkunan koon muutokset. myös max window
> history back on tärkeä. history popup myös.
> 
> studypad edit moodissa "new item" tärkeä.
> Study pad edit moodissa hyppiminen edelliseen editoitavaan ja seuraavaan tärkeä. 
> 
> bible selectorissa pitäs keyboardillakin lähtöpiste olla se mikä on highlightattu, mutta ei ole.
> 
> jump to verse bibleviewin puolella on selkeä ja nopea toteuttaa.
> 
> Jonkinlainen "find" myös tärkeä. Ja tärkeä että se ei turmele historiaa

.
